### PR TITLE
Add EventAggregator

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes 

Touches #9397 and #8653 

### Description

Introduces a simple `IEventAggregator` implementation based upon Mediatr.

This allows the registration of `INotificationHandler<TNotification>` types against an `IUmbracoBuilder`.

Unlimited handlers can be registered.
```c#
builder.AddNotificationHandler<Notification, NotificationHandlerA>();
builder.AddNotificationHandler<Notification, NotificationHandlerB>();
builder.AddNotificationHandler<Notification, NotificationHandlerC>();
```

Triggering events is as follows.

```c#
var notification = new Notification();
await aggregator.PublishAsync(notification);
```

The aggregator can be used eventually to replace all static events . `IComponent` types can also be converted into handlers that subscribe to specific Umbraco events. In addition, plugins like UmbracoForms can provide their own notification events with which a developer could subscribe.

I've left scope for allowing the development of an API for the registration of individual request handlers `IRequestHandler<TRequest>` of which there can only ever be one but not implemented it at this point.

\CC @rustybox @lars-erik

### CODE REVIEW

**Comment by rustybox** [comment]
What's the theory on using the async api from say ContentService, lots of GetAwaiter().GetResult?

**Comment by rustybox** [comment]
> IRequestHandler

Off topic, I do wish that was split out to ICommandHandler & IQueryHandler in mediatr.

> Mocking should be an act of desperation!

Off topic, I take issue with this statement in the general sense, great way to assert methods called/not called by a SUT on a service without risking side effects.

**Comment by JimBobSquarePants** [comment]
>What's the theory on using the async api from say ContentService, lots of GetAwaiter().GetResult?

`ContentService` should be rewritten to be asynchronous. Everything should be otherwise you risk deadlocks. 

>Off topic, I take issue with this statement in the general sense, great way to assert methods called/not called by a SUT on a service without risking side effects.

Except in this case there are side effects caused by an incomplete implementation. 

Mock objects have their use, but when not used carefully, they often encourage bad practices, testing implementation details, hinder refactoring and produce difficult to read and difficult to maintain tests. Here's a great article from Oracle listing the shortcomings of using mocks. 

https://www.oracle.com/technical-resources/articles/enterprise-architecture/mock-shortcomings.html

**Comment by JimBobSquarePants** [comment]
>Off topic, I do wish that was split out to ICommandHandler & IQueryHandler in mediatr.

Yeah, I'd agree here. It would be much more clear - I don't see why we couldn't do that.

**Comment by rustybox** [comment]
> Except in this case there are side effects caused by an incomplete implementation.

Indeed, that's what I meant by in the general sense, you're statement wasn't qualified to this instance.

In this case shit's weird, why even have that method in the first place, new it up per test, the helper call is the same length as the instantiation or the mock call.

**Comment by rustybox** [comment]
> ContentService should be rewritten to be asynchronous. Everything should be otherwise you risk deadlocks.

Hah I thought you'd say that, that'sa whole lot of refactoring :D

**Comment by JimBobSquarePants** [comment]
I'm gonna push a further commit that's going to add some noise to the PR. 

It normalizes the `IUmbracoBuilder`, `UmbracoBuilder` types to use the same namespace `Umbraco.Core.DependencyInjection`.

`I[Umbraco]Composer` should be there too but I'll leave that for now.

**Comment by stevetemple** [comment]
I like it, nice and simple

**Comment by lars-erik** [comment]
> Off topic, I do wish that was split out to ICommandHandler & IQueryHandler in mediatr

I think those are three different things. Events are something else than commands and queries.  
Granted, "examine is done" could be handled by chains of "do the examine thing" command handlers.  
But I'd still like for the examine domain to publish a notification to subscribers that _something happened_, which isn't necessarily a command or a query.  
The latter two are often specific user actions IMO, while events are stuff happening in the system as smaller and more diverse concequences of those.  

This is *not* to say that I don't _also_ want ICommandHandler and IQueryHandler with respective base messages. 🤓👍

**Comment by rustybox** [comment]
Events is INotificationHandler

 ICommandHandler + IQueryHandler = IRequestHandler

**Comment by lars-erik** [comment]
Reg. the "only ever one" handler - I don't have a problem with that as long as it can be decorated / composite. 👼  
Die on the hill of three different concepts and hence three different contract/message abstractions. 👍

**Comment by rustybox** [comment]
yep three things, 3 contracts, of which we only need 1 to replace the static events :)

**Comment by JimBobSquarePants** [comment]
Query/Command can be dealt with separately. I’m just keen to see this reviewed/merged so we can continue to clean up.

**Comment by bergmania** [comment]
> Query/Command can be dealt with separately. I’m just keen to see this reviewed/merged so we can continue to clean up.

I know, but I really don't have time to do a proper review this week 😢

**Comment by JimBobSquarePants** [comment]
>I know, but I really don't have time to do a proper review this week 😢

Trust me, it's perfect 😜

**Review by bergmania — COMMENTED** [review]
Hi @JimBobSquarePants .. 

Super PR.. 
Just a few comments. Could as well be fixed in upcoming PRs.. In most cases it is only duplicate usings where we have multiple other necessary usings..

File: src/Umbraco.Tests.Integration/RuntimeTests.cs [hunk 1]
```diff
@@ -23,7 +23,7 @@
 using Umbraco.Tests.Integration.Extensions;
 using Umbraco.Tests.Integration.Implementations;
 using Umbraco.Tests.Integration.Testing;
-using Umbraco.Web.Common.Builder;
+using Umbraco.Core.DependencyInjection;
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/Umbraco.Tests.Integration/RuntimeTests.cs]: Fixed

File: src/Umbraco.Tests.Integration/TestServerTest/UmbracoBuilderExtensions.cs [hunk 2]
```diff
@@ -1,11 +1,11 @@
 ﻿using Moq;
 using Umbraco.Core;
-using Umbraco.Core.Builder;
+using Umbraco.Core.DependencyInjection;
 using Umbraco.Core.Cache;
 using Umbraco.Core.Logging;
 using Umbraco.Core.Runtime;
 using Umbraco.Tests.Integration.Implementations;
-using Umbraco.Web.Common.Builder;
+using Umbraco.Core.DependencyInjection;
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/Umbraco.Tests.Integration/TestServerTest/UmbracoBuilderExtensions.cs]: Fixed

File: src/Umbraco.Tests.UnitTests/Umbraco.Core/Components/ComponentTests.cs [hunk 3]
```diff
@@ -21,7 +21,7 @@
 using Umbraco.Core.Persistence.Mappers;
 using Umbraco.Core.Scoping;
 using Umbraco.Tests.TestHelpers;
-using Umbraco.Web.Common.Builder;
+using Umbraco.Core.DependencyInjection;
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/Umbraco.Tests.UnitTests/Umbraco.Core/Components/ComponentTests.cs]: Fixed

File: src/Umbraco.Web.BackOffice/Extensions/UmbracoBuilderExtensions.cs [hunk 4]
```diff
@@ -2,10 +2,10 @@
 using Microsoft.AspNetCore.Mvc.Filters;
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.VisualBasic;
-using Umbraco.Core.Builder;
+using Umbraco.Core.DependencyInjection;
 using Umbraco.Web.BackOffice.Filters;
 using Umbraco.Web.BackOffice.Security;
-using Umbraco.Web.Common.Builder;
+using Umbraco.Core.DependencyInjection;
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/Umbraco.Web.BackOffice/Extensions/UmbracoBuilderExtensions.cs]: Fixed

File: src/Umbraco.Web.Common/Builder/UmbracoBuilderExtensions.cs [hunk 5]
```diff
@@ -20,7 +20,7 @@
 using Smidge;
 using Smidge.Nuglify;
 using Umbraco.Core;
-using Umbraco.Core.Builder;
+using Umbraco.Core.DependencyInjection;
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/Umbraco.Web.Common/Builder/UmbracoBuilderExtensions.cs]: Fixed

File: src/Umbraco.Web.Website/Extensions/UmbracoBuilderExtensions.cs [hunk 6]
```diff
@@ -1,8 +1,8 @@
 ﻿using Microsoft.AspNetCore.Mvc;
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Options;
-using Umbraco.Core.Builder;
-using Umbraco.Web.Common.Builder;
+using Umbraco.Core.DependencyInjection;
+using Umbraco.Core.DependencyInjection;
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/Umbraco.Web.Website/Extensions/UmbracoBuilderExtensions.cs:5]: Fixed

**Comment by JimBobSquarePants** [comment]
@bergmania I've updated everything to use new rules 😍 and also fixed the duplicate namespaces.

File: src/Umbraco.Tests.UnitTests/Umbraco.Core/Composing/CollectionBuildersTests.cs [hunk 7]
```diff
@@ -6,13 +6,13 @@
...
 using Umbraco.Core;
-using Umbraco.Core.Builder;
+using Umbraco.Core.DependencyInjection;
 using Umbraco.Core.Cache;
 using Umbraco.Core.Composing;
 using Umbraco.Core.Logging;
 using Umbraco.Tests.TestHelpers;
 using Umbraco.Tests.UnitTests.TestHelpers;
-using Umbraco.Web.Common.Builder;
+using Umbraco.Core.DependencyInjection;
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/Umbraco.Tests.UnitTests/Umbraco.Core/Composing/CollectionBuildersTests.cs]: Fixed

File: src/Umbraco.Tests.Integration/Testing/UmbracoIntegrationTest.cs [hunk 8]
```diff
@@ -33,8 +33,8 @@
 using Microsoft.Extensions.Logging;
 using Microsoft.Extensions.Logging.Abstractions;
 using Serilog;
-using Umbraco.Core.Builder;
-using Umbraco.Web.Common.Builder;
+using Umbraco.Core.DependencyInjection;
+using Umbraco.Core.DependencyInjection;
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/Umbraco.Tests.Integration/Testing/UmbracoIntegrationTest.cs]: Fixed

**Comment by bergmania** [comment]
Good work @JimBobSquarePants

**Comment by lars-erik** [comment]
This might be the start of beautiful thing! 😍

**Pull request merged**